processing/src/main/java/org/apache/druid/query/filter/ArrayContainsFilter.java

```diff
@@ -0,0 +1,677 @@
+      // column is not an array, behave like a normal equality filter
+      final ValueIndexes valueIndexes = indexSupplier.as(ValueIndexes.class);
+      if (valueIndexes != null) {
+        // matchValueEval.value() cannot be null here due to check in the constructor
+        //noinspection DataFlowIssue
+        return valueIndexes.forValue(elementMatchValueEval.value(), elementMatchValueType);
+      }
+
+      if (elementMatchValueType.isPrimitive()) {
+        final StringValueSetIndexes stringValueSetIndexes = indexSupplier.as(StringValueSetIndexes.class);
+        if (stringValueSetIndexes != null) {
+
+          return stringValueSetIndexes.forValue(elementMatchValueEval.asString());
+        }
+      }
+    }
+    // column exists, but has no indexes we can use
+    return null;
+  }
+
+  @Override
+  public ValueMatcher makeMatcher(ColumnSelectorFactory factory)
+  {
+    return ColumnProcessors.makeProcessor(
+        column,
+        new TypedConstantElementValueMatcherFactory(elementMatchValueEval, predicateFactory),
+        factory
+    );
+  }
+
+  @Override
+  public VectorValueMatcher makeVectorMatcher(VectorColumnSelectorFactory factory)
+  {
+    final ColumnCapabilities capabilities = factory.getColumnCapabilities(column);
+
+    if (elementMatchValueType.isPrimitive() && (capabilities == null || capabilities.isPrimitive())) {
+      return ColumnProcessors.makeVectorProcessor(
+          column,
+          VectorValueMatcherColumnProcessorFactory.instance(),
+          factory
+      ).makeMatcher(elementMatchValueEval.value(), elementMatchValueType);
+    }
+    return ColumnProcessors.makeVectorProcessor(
+        column,
+        VectorValueMatcherColumnProcessorFactory.instance(),
+        factory
+    ).makeMatcher(new ArrayContainsPredicateFactory(elementMatchValueEval));
+  }
+
+  @Override
+  public boolean supportsSelectivityEstimation(ColumnSelector columnSelector, ColumnIndexSelector indexSelector)
+  {
+    return Filters.supportsSelectivityEstimation(this, column, columnSelector, indexSelector);
+  }
+
+  @Override
+  public boolean canVectorizeMatcher(ColumnInspector inspector)
+  {
+    return true;
+  }
+
+  @Override
+  public Set<String> getRequiredColumns()
+  {
+    return ImmutableSet.of(column);
+  }
+
+  @Override
+  public boolean supportsRequiredColumnRewrite()
+  {
+    return true;
+  }
+
+  @Override
+  public Filter rewriteRequiredColumns(Map<String, String> columnRewrites)
+  {
+    String rewriteDimensionTo = columnRewrites.get(column);
+
+    if (rewriteDimensionTo == null) {
+      throw new IAE(
+          "Received a non-applicable rewrite: %s, filter's dimension: %s",
+          columnRewrites,
+          columnRewrites
+      );
+    }
+
+    return new EqualityFilter(
+        rewriteDimensionTo,
+        elementMatchValueType,
+        elementMatchValue,
+        filterTuning
+    );
+  }
+
+  private static class ArrayContainsPredicateFactory implements DruidPredicateFactory
```

Comment:
Looks like a lot of copypaste from `EqualityFilter` in the non-array cases. Would it make sense to delegate to `EqualityFilter` or use common helpers instead?